Make active power and power meter readings as close as possible so computing consumption values are more accurate

JacoboSanchez opened the pull request:
As power meter provides a variation over the active power and both values should be combined in order to compute the power usage they should be obtained as close in time as possible. As they are in different memory sections it is not possible to retrieve them in a single modbus read so reading it in the following line seems the best workaround to prevent invalid power usage values.

I have other modifications in my fork like providing a prefix option in case you have two inverters of the same model you can prevent their entity name to conflict, i can make a PR for that if you think it may be useful. I have also a reduced attributes option to minimize modbus calls but that is based on my personal needs.

I am also considering passing the power usage and self consumption computations to attributes to the extension instead of making the active power and power meter active power substraction as sensor templates in home assistant but i am not convinced that adding derived values here is the way to go... but it will make users to not care abouut how they should be computed from the data returned by the inverter... what's your opinion?

Emilv2 commented:
First of all thank you for your contribution!

> As power meter provides a variation over the active power and both values should be combined in order to compute the power usage they should be obtained as close in time as possible. As they are in different memory sections it is not possible to retrieve them in a single modbus read so reading it in the following line seems the best workaround to prevent invalid power usage values.

Looks good to me, do you have any issues with values being incorrect otherwise?

> I have other modifications in my fork like providing a prefix option in case you have two inverters of the same model you can prevent their entity name to conflict, i can make a PR for that if you think it may be useful.

My plan is to get the component officially into HA, for that I need to convert it to config flow and I'm not really sure how that would fit together, but if it's not a too big change I can add it for now.

> I have also a reduced attributes option to minimize modbus calls but that is based on my personal needs.

I'm trying to keep new changes as close as possible to HA requirements to make the transition not too hard and I don't think they like too many config options to retrieve extra data (but I could be wrong). Do you have any problems caused by too many modbus calls? If there's not really an issue I don't think we need to make things more complicated than necessary. Otherwise I think it could be useful for other people who might have the same problems.

> I am also considering passing the power usage and self consumption computations to attributes to the extension instead of making the active power and power meter active power substraction as sensor templates in home assistant but i am not convinced that adding derived values here is the way to go... but it will make users to not care abouut how they should be computed from the data returned by the inverter... what's your opinion?

Same thing here, I want to follow official HA guidelines for new sensors. I already know many of the attributes should become their own sensor, but I'm not really sure about derived values. Personally I think adding it is fine.

Starting next month I'll have some time again to focus on this more, you're welcome to help! Especially the porting to config flow, I'm not really looking forward to that.

JacoboSanchez commented:
I think you are doing it wright by limiting the config options, that's why i just did the PR for the minor code update. However as i was making modifications on your project i feel that at least i should let you know just in case. For the prefix option i don't even need it as i have two inverters but they are 2k and 3k so there is no conflict... just trying if i could have fixed in case they were. 

I am mainly a Java developer and still don't feel comfortable with python so use home assistant to think about some minor change and try to dig into code to understand how it works (same with JS in some custom cards).

I had some charts where i compare production/consumption/self consumption and on partially cloudy days (with a high production variation) i started to get negative values on consumption values (active power minus power meter) so i added checks for negative values on sensor templates.  For example for self consumption:

` value_template: >-
          {% set t = states('sensor.sun2000_2ktl_l1') | int %}
          {% set u = state_attr('sensor.sun2000_2ktl_l1', 'power_meter_active_power') | int %}
          {% if u > t %} <====== this should never happen
              {{ (0) | int }}
          {% elif u > 0 %}
              {{ (t - u) }}
          {% else %}
              {{ (t) }}
          {% endif %}`

Then made a test by retrieving another active power measure at the end of the update code and saw that their value was different sometimes.
At first tried to reduce calls by calling directly the inverter in memory groups instead of single records... but that was too much work for me and the more important values were not in the same memory range so i gave up. Then i add the negative checks and did the reading modification from this PR to minimize it (i also added an option to minimize calls by removing attributes i am not currently using but that was to prove me i understand the code)

Nothing more, thanks for your work and keep it up!

Chart example using derived values (currently there are no negatives as i transform them into 0s):
![image](https://user-images.githubusercontent.com/11249890/118378596-3cf0af00-b5d5-11eb-83ca-bc16853a5073.png)
